src/object/obj_tx.c

```diff
@@ -314,7 +311,9 @@ dc_tx_alloc(daos_handle_t coh, daos_epoch_t epoch, uint64_t flags,
 			tx->tx_epoch.oe_flags = DTX_EPOCH_UNCERTAIN;
 		}
 	} else {
+		daos_dti_gen(&tx->tx_id, true);
```

Comment:
Why don't we generate a TX ID in this case? Because a TX ID will not be used? This PR also removes the RDONLY flag from the "local" dc_tx_alloc calls. That seems to imply those "local" cases might want to write in the future. Don't they need TX IDs if they write?

Reply:
If the logic is not easy to clarify, I'd vote for just generating a TX ID in all cases.

Reply:
Because we do not want to leave read TS on server for local transaction or the snapshot based transaction.

The patch does not remove "READONLY" flags for local transaction, instead, we force local transaction must be "READONLY". Local transaction has specified epoch, we will not generate DTX ID for it, then it will neither leave read TS on server nor be required to be restarted because of MVCC conflict.

As for further possibility of "write" via local transaction, I am not sure whether we will have such use cases. If we have to support that, we can adjust related logic when need at that time. It will affect neither RPC nor TX API.

Reply:
> Because we do not want to leave read TS on server for local transaction or the snapshot based transaction.

OK, I recall.

> The patch does not remove "READONLY" flags for local transaction, instead, we force local transaction must be "READONLY".

Indeed. I missed this change.